### PR TITLE
Adds the endpoint that triggers data transfer 

### PR DESCRIPTION
**Pull request by josenavas**
~~Depends on #214 - so review merge that one first~~

The code generating the prep information was done by @wasade

### CODE REVIEW

File: knimin/lib/qiita_jira_util.py [hunk 1]
```diff
@@ -345,20 +313,66 @@ def create_sequencing_run(pool_id, email, sequencer, reagent_type,
...
+        _update_qiita_samples(study_id, blanks, replicates)
+
+    # Write the sample sheet
+    instrument_type = 'miseq' if instrument_model == 'MiSeq' else 'hiseq'
+    run_type = "Target Gene" if pool['targeted_pools'] else "Shotgun"
+    sample_information = None
+
+    # TODO: Fill sample_information if the run is a HiSeq
+    # TODO: I want to check with Greg the values for PI and contact
+    contact_0_name = contact_0_email = pi_email = pi_name = 'TODO'
```
Comment (wasade) [inline knimin/lib/qiita_jira_util.py]: Should these be fetched from the db right now?

File: knimin/lib/qiita_jira_util.py [hunk 2]
```diff
@@ -367,13 +381,163 @@ def create_sequencing_run(pool_id, email, sequencer, reagent_type,
...
+            # Copy the sequencing files to relevant Qiita folders
+            atype, filepaths = _copy_sequence_files_to_qiita(prep, run_path)
+
+            dtype = (prep.target_gene.iloc[0]
+                     if 'target_gene' in prep else 'Metagenomics')
+
+            prep, _ = _assess_replicates(prep)
+            prep['created_on'] = prep.created_on.apply(lambda x: x.isoformat())
+            # Cast everything to string: some types generate issues
+            prep = prep.astype(str)
```
Comment (wasade) [inline knimin/lib/qiita_jira_util.py]: is this already enforced in the `generate_prep_information` step, and if not, should it be? super nervous about the implicit casts...

File: knimin/lib/data_access.py [hunk 3]
```diff
@@ -4904,6 +4905,119 @@ def delete_sequencing_run(self, run_id):
...
+        with TRN:
+            targeted_preps = []
+            for targeted_pool in pool['targeted_pools']:
+                TRN.add(tg_sql, [targeted_pool['targeted_plate_id']])
+                for results in TRN.execute_fetchindex():
+                    results = dict(results)
+                    results.update(row_updater)
+                    targeted_preps.append(results)
+
+            # shotgun_preps = []
```
Comment (wasade) [inline knimin/lib/data_access.py]: should these be removed for right now?

File: knimin/lib/data_access.py [hunk 4]
```diff
@@ -4904,6 +4905,119 @@ def delete_sequencing_run(self, run_id):
...
+            #         shotgun_preps.append(results)
+
+        target_prep = pd.DataFrame(targeted_preps)
+        # shotgun_prep = pd.DataFrame(shotgun_preps)
+
+        preps = []
+        for fullprep in [target_prep]:  # , shotgun_prep]:
+            blank = fullprep.loc[fullprep.is_blank]
+            notblank = fullprep.loc[~fullprep.is_blank]
+            notblank.loc[:, 'study_id'] = notblank.study_id.astype(int)
```
Comment (ElDeveloper) [inline knimin/lib/data_access.py]: Can this be removed assuming this is already an integer?

File: knimin/lib/tests/test_data_access.py [hunk 5]
```diff
@@ -2142,6 +2143,171 @@ def test_delete_sequencing_run(self):
...
+        run_id = db.create_sequencing_run(
+            pool_id, 'test', None, 'MiSeq v3 150 cycle', 'MS1234',
+            'Illumina', 'MiSeq', 'TrueSeq HT', 151, 151)
+        self._clean_up_funcs.insert(
+            0, partial(db.delete_sequencing_run, run_id))
+
+        # Generate the prep information
+        preps = db.generate_prep_information(run_id)
+
+        # There where 2 studies in this run, so we should get 2 preps
```
Comment (ElDeveloper) [inline knimin/lib/tests/test_data_access.py]: where -> were

File: knimin/lib/qiita_jira_util.py [hunk 6]
```diff
@@ -133,6 +129,43 @@ def _create_kl_jira_project(jira_user, jira_template, study_id, title):
...
+
+    Parameters
+    ----------
+    prep : pandas.DataFrame
+        The prep information
+
+    Returns
+    -------
+    pandas.DataFrame, list of str
+        The updates pandas dataframe and a list of replicates
```
Comment (ElDeveloper) [inline knimin/lib/qiita_jira_util.py]: updates -> updated

File: knimin/lib/qiita_jira_util.py [hunk 7]
```diff
@@ -133,6 +129,43 @@ def _create_kl_jira_project(jira_user, jira_template, study_id, title):
...
+    """
+    counts = prep.sample_name.value_counts()
+    replicates = counts[counts > 1].index.tolist()
+    # Rename the samples by adding the plate id and the well.
+    # Rename the column sample_name
+    prep.rename(columns={'sample_name': 'original_sample_name'},
+                inplace=True)
+    # Create the new sample name column
+    prep['sample_name'] = prep['original_sample_name']
```
Comment (ElDeveloper) [inline knimin/lib/qiita_jira_util.py]: Minor, but the prep.rename + prep['sample_name'] ... could be changed to:

prep['original_sample_name'] = prep['sample_name'].copy()

File: knimin/lib/qiita_jira_util.py [hunk 8]
```diff
@@ -133,6 +129,43 @@ def _create_kl_jira_project(jira_user, jira_template, study_id, title):
...
+    for r in replicates:
+        for i in prep[prep['original_sample_name'] == r].index:
+            sample_id = _format_sample_id(
+                prep.original_sample_name[i], prep.dna_plate_id[i],
+                prep.row[i] - 1, prep.col[i] - 1)
+            prep.loc[i, 'sample_name'] = sample_id
+            new_sample_ids.append(sample_id)
+
+    # It is ensured that the sample_name column contain unique IDs
+    prep.set_index('sample_name', inplace=True, drop=True)
```
Comment (ElDeveloper) [inline knimin/lib/qiita_jira_util.py]: If the goal is to check that no sample_names are repeated, then you can add the `verify_integrity` argument to set_index, see here:

http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.set_index.html

File: knimin/lib/qiita_jira_util.py [hunk 9]
```diff
@@ -378,6 +344,18 @@ def create_sequencing_run(pool_id, email, reagent_type, reagent_lot, platform,
     run = db.read_sequencing_run(run_id)
     pool = db.read_pool(run['pool_id'])
 
+    # To make sure that we generate the sample sheet correctly, we need to
+    # assess the technical replicates at this time
+    preps = db.generate_prep_information(run_id)
+    for prep in preps:
+        prep, replicates = _assess_replicates(prep)
+        # Update qiita with blanks and replicates
+        blanks = [sid for sid in prep[prep.is_blank].index]
```
Comment (ElDeveloper) [inline knimin/lib/qiita_jira_util.py]: This is the same as `prep[prep.is_blank].index.tolist()` right?

File: knimin/lib/qiita_jira_util.py (hunk 2 above)
Comment (ElDeveloper) [inline knimin/lib/qiita_jira_util.py]: This shouldn't be an issue as all the data in here should be Python types.

File: knimin/lib/data_access.py (hunk 4 above)
Comment (josenavas) [inline knimin/lib/data_access.py]: Done

File: knimin/lib/qiita_jira_util.py (hunk 6 above)
Comment (josenavas) [inline knimin/lib/qiita_jira_util.py]: Done

File: knimin/lib/qiita_jira_util.py (hunk 7 above)
Comment (josenavas) [inline knimin/lib/qiita_jira_util.py]: Done

File: knimin/lib/qiita_jira_util.py (hunk 8 above)
Comment (josenavas) [inline knimin/lib/qiita_jira_util.py]: Thanks for the link! The check for duplicates was done for knowing how to rename the samples accordingly for technical replicates. What the comment says is that at this point I'm sure there are no replicates cause I would've rename it in the previous for loop.

File: knimin/lib/qiita_jira_util.py (hunk 9 above)
Comment (josenavas) [inline knimin/lib/qiita_jira_util.py]: Good call

File: knimin/lib/tests/test_data_access.py (hunk 5 above)
Comment (josenavas) [inline knimin/lib/tests/test_data_access.py]: Done

**Comment by josenavas** [comment]
While the tests are running, can I get another round of reviews @ElDeveloper and @wasade ? unless you already agree with the changes.

**Comment by wasade** [comment]
Reviewing

**Comment by wasade** [comment]
:+1:

**Comment by josenavas** [comment]
Yay! Tests passing again!
@wasade wanna do the honors?

**Pull request merged**